### PR TITLE
feat(`forge build -vvvvv`): If verbosity level is 5 or higher show files to compile

### LINKED ISSUE #7888: feat(`forge build`): add flag to show names of files that are not cached and being compiled
### Component

Forge

### Describe the feature you would like

We have a ci workflow that downloads cached foundry build output before running `forge build`, in order to speed up compilation time.  Sometimes foundry recompiles some of the files, even though it seems that the source code hasn't changed between builds.  It would be great to add a flag that shows the names of the files being compiled.

For example, running `forge build`, when there are 221 files, and only some of the files have changed since the last build, produces the output while it compiles:
`Compiling 20 files with 0.8.17`

It would be great to be able to know the names of those 20 files.

If you add the `--names` flag, it will show all 221 files, regardless of if they were compiled or laoded from cache.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mgiagante**
Closes #7888 

This PR is to address [Issue 7888](https://github.com/foundry-rs/foundry/issues/7888) by leveraging the new global verbosity level flag implemented by https://github.com/foundry-rs/foundry/pull/9273

## Motivation
This aims to provide users with a way to see which files have changes relative to what's cached, and thus are effectively picked up by the compiler when running `forge build`.

## Solution
If `forge build -v` is run, this will set the verbosity level to 1. A level of 1 or higher (with additional v's)  will display the list of files being compiled.

### CODE REVIEW

**Review by grandizzy — CHANGES_REQUESTED** [review]
thank you! I think this could also close https://github.com/foundry-rs/foundry/issues/9304 so when passing `-v` flag avoids confusion re files that are compiled, @zerosnacks wdyt?

**Comment by zerosnacks** [comment]
> thank you! I think this could also close #9304 so when passing `-v` flag avoids confusion re files that are compiled, @zerosnacks wdyt?

Supportive, I think a warning here would still be beneficial as well as this `-v` flag would indicate the files being compiled but that doesn't offer an explanation as to why they are. I think raising a warning on the fallback trigger (if a path is specified) would still be the right way but this PR is in the right direction

File: crates/common/src/term.rs
Comment (mgiagante) [reply crates/common/src/term.rs:163]: Awesome! I'll go with this. Thanks!

Comment (mgiagante) [reply crates/common/src/term.rs:159]: @grandizzy after implementing it this way, there's still some interference between the spinner and the `dirty_files` list.
```
/home/mg/Code/foundry-test-project/lib/forge-std/src/safeconsole.sol
/home/mg/Code/foundry-test-project/lib/forge-std/src/StdUtils.sol
/home/mg/Code/foundry-test-project/script/Counter.s.sol
/home/mg/Code/foundry-test-project/lib/forge-std/src/StdError.sol
[⠑] Compiling filesCode/foundry-test-project/lib/forge-std/src/StdCheats.sol
/home/mg/Code/foundry-test-project/lib/forge-std/src/console.sol
/home/mg/Code/foundry-test-project/lib/forge-std/src/StdChains.sol
/home/mg/Code/foundry-test-project/lib/forge-std/src/StdMath.sol
/home/mg/Code/foundry-test-project/lib/forge-std/src/Vm.sol
```

The spinner displayed by the `SpinnerReporter` shows up in the middle of the file list.
This seems to be because `SpinnerReporter::spawn()` is called before the compilation process starts, in `crates/common/src/compile.rs:477`:

```
/// Configures the reporter and runs the given closure.
pub fn with_compilation_reporter<O>(quiet: bool, f: impl FnOnce() -> O) -> O {
    #[allow(clippy::collapsible_else_if)]
    let reporter = if quiet {
        Report::new(NoReporter::default())
    } else {
        if std::io::stdout().is_terminal() {
            Report::new(SpinnerReporter::spawn())
        } else {
            Report::new(BasicStdoutReporter::default())
        }
    };

    foundry_compilers::report::with_scoped(&reporter, f)
}
```

The hook in which the `dirty_files` are available to be printed out is `on_compiler_spawn` in `crates/common/src/term.rs:158` and it only runs after the spinner is spawned.

Comment (grandizzy) [reply crates/common/src/term.rs:159]: There shouldn't be an if / else, just if verbose print dirty files. The other should always be printed when compiler spawned.

Comment (mgiagante) [reply crates/common/src/term.rs:159]: `on_compiler_spawn` looks like the best place to output the dirty files list, but since it necessarily runs after the spinner reporter is spawned I'm wondering if it makes sense doing it somewhere else.

Comment (mgiagante) [reply crates/common/src/term.rs:168]: > yeah, what I meant is that this else should be removed and always send 2nd message, regardless verbosity
> 
> https://github.com/foundry-rs/foundry/blob/c8a3362169891c575a674ff31b43e909b9e17b72/crates/common/src/term.rs#L168

Got it, thanks. This one's been addressed.

Comment (mgiagante) [reply crates/common/src/term.rs:168]: :thinking: I think the message interfering with it is the `[⠑] Compiling files` displayed for the spinner every time it ticks. I still see it happen after removing the `if..else`:
```
/home/mg/Code/mgiagante/foundry-test-project/lib/forge-std/src/interfaces/IERC20.sol
[⠑] Compiling filesante/foundry-test-project/lib/forge-std/src/interfaces/IMulticall3.sol
/home/mg/Code/mgiagante/foundry-test-project/script/Counter.s.sol
```

Comment (grandizzy) [reply crates/common/src/term.rs:168]: the verbose message
https://github.com/foundry-rs/foundry/blob/6b64f2b3864d6e7addbb566a4b7df88085823ca6/crates/common/src/term.rs#L168-L178
should be sent first, like

```rust
        if foundry_common::shell::verbosity() > 0 {
            self.send_msg(format!(
                "Compiling files\n{}",
                dirty_files
                    .iter()
                    .map(|path| path.display().to_string())
                    .collect::<Vec<String>>()
                    .join("\n")
            ));
        }
        self.send_msg(format!(
            "Compiling {} files with {} {}.{}.{}",
            dirty_files.len(),
            compiler_name,
            version.major,
            version.minor,
            version.patch
        ));
```

**Comment by mgiagante** [comment]
@zerosnacks @grandizzy is there currently a way to write tests for this? I've tried the tests in `crates/forge/tests/cli/build.rs` but the output that needs to be asserted on here doesn't seem to be included.

**Comment by grandizzy** [comment]
> build.rs

@mgiagante I think that's fine without a test, sorry for back and forth, could you remove the newly added variant `SpinnerMsg::VerboseMsg` (which is not really required) and use same `SpinnerMsg::VerboseMsg` / `send_msg` for sending verbose message, then this should be good to send. thank you

**Review by grandizzy — APPROVED** [review]
thank you! pending one more review before merging

**Review by zerosnacks — APPROVED** [review]
Made some minor modifications to the rendered output

Now renders as (path trimmed, sorted, prefixed):

```
[⠊] Compiling...
[⠒] Files to compile:
- lib/forge-std/src/Base.sol
- lib/forge-std/src/Script.sol
- lib/forge-std/src/StdAssertions.sol
- lib/forge-std/src/StdChains.sol
- lib/forge-std/src/StdCheats.sol
- lib/forge-std/src/StdError.sol
- lib/forge-std/src/StdInvariant.sol
- lib/forge-std/src/StdJson.sol
- lib/forge-std/src/StdMath.sol
- lib/forge-std/src/StdStorage.sol
- lib/forge-std/src/StdStyle.sol
- lib/forge-std/src/StdToml.sol
- lib/forge-std/src/StdUtils.sol
- lib/forge-std/src/Test.sol
- lib/forge-std/src/Vm.sol
- lib/forge-std/src/console.sol
- lib/forge-std/src/console2.sol
- lib/forge-std/src/interfaces/IERC165.sol
- lib/forge-std/src/interfaces/IERC20.sol
- lib/forge-std/src/interfaces/IERC721.sol
- lib/forge-std/src/interfaces/IMulticall3.sol
- lib/forge-std/src/mocks/MockERC20.sol
- lib/forge-std/src/mocks/MockERC721.sol
- lib/forge-std/src/safeconsole.sol
- script/Counter.s.sol
- src/Counter.sol
- src/Counter2.sol
- src/Counter3.sol
- src/Counter4.sol
- test/Counter.t.sol
[⠒] Compiling 30 files with Solc 0.8.28
[⠆] Solc 0.8.28 finished in 950.06ms
Compiler run successful!
```

As opposed to

```
[⠒] Compiling files
[⠊] Compiling...
[⠒] Compiling files
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdChains.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/interfaces/IERC165.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/Test.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/interfaces/IERC20.sol
/home/zerosnacks/Projects/counter/src/Counter.sol
/home/zerosnacks/Projects/counter/src/Counter3.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/interfaces/IERC721.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdStyle.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdToml.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/mocks/MockERC721.sol
/home/zerosnacks/Projects/counter/test/Counter.t.sol
/home/zerosnacks/Projects/counter/src/Counter2.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/Base.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/mocks/MockERC20.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdError.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdMath.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdCheats.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/Vm.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdInvariant.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/Script.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/interfaces/IMulticall3.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/console2.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/console.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdStorage.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdAssertions.sol
/home/zerosnacks/Projects/counter/src/Counter4.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/safeconsole.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdJson.sol
/home/zerosnacks/Projects/counter/lib/forge-std/src/StdUtils.sol
/home/zerosnacks/Projects/counter/script/Counter.s.sol
[⠢] Compiling 30 files with Solc 0.8.28
[⠆] Solc 0.8.28 finished in 984.73ms
Compiler run successful
```

cc @grandizzy

**Comment by DaniPopes** [comment]
This is extremely verbose, and it is common for verbosity to be set to 2 or 3 when running `test`. Unfortunately coupling testing verbosity to everywhere else means that we need to set higher verbosity values for stuff like this.

I'd argue this should be like a trace level log, so 5 or above

File: crates/common/src/term.rs
Comment (DaniPopes) [inline crates/common/src/term.rs:173]:
```suggestion
                    .format("\n")
```

**Comment by mgiagante** [comment]
> This is extremely verbose, and it is common for verbosity to be set to 2 or 3 when running `test`. Unfortunately coupling testing verbosity to everywhere else means that we need to set higher verbosity values for stuff like this.
> 
> I'd argue this should be like a trace level log, so 5 or above

That makes a lot of sense. I'll go with 5 then.
Thanks!

**Comment by zerosnacks** [comment]
> This is extremely verbose, and it is common for verbosity to be set to 2 or 3 when running `test`. Unfortunately coupling testing verbosity to everywhere else means that we need to set higher verbosity values for stuff like this.
> 
> I'd argue this should be like a trace level log, so 5 or above

Hmm.. yeah this is not great. For now let's do `-vvvvv` to avoid interference.

Ideally we would have a way to distinguish if the user intentionally runs `forge build` or it is part of a process like `forge test` and adapt based on that. I suggest post-V1 we consider how to apply a more uniform scale.

**Comment by mgiagante** [comment]
> > This is extremely verbose, and it is common for verbosity to be set to 2 or 3 when running `test`. Unfortunately coupling testing verbosity to everywhere else means that we need to set higher verbosity values for stuff like this.
> > I'd argue this should be like a trace level log, so 5 or above
> 
> Hmm.. yeah this is not great. For now let's do `-vvvvv` to avoid interference.
> 
> Ideally we would have a way to distinguish if the user intentionally runs `forge build` or it is part of a process like `forge test` and adapt based on that. I suggest post-V1 we consider how to apply a more uniform scale.

Just created https://github.com/foundry-rs/foundry/issues/9340 to track this.

**Comment by zerosnacks** [comment]
> Just created #9340 to track this.

All good, added it as a note in https://github.com/foundry-rs/foundry/issues/8794

**Pull request merged**
